### DISCUSSION TITLE
Recommended way to deal with environment variables in a `nx-monorepo`

### DISCUSSION THREAD

**Discussion by zetashift** [aws/aws-pdk#404] [Q&A] [answered]
I was wondering what the recommended way is to deal with environment variables here is.

Some of my environment variables include:
- Database DSN, for connecting to a database (this one gets injected in the Lambdas env variables)
- Access tokens
- Lambda API client URL

Do I create a subproject, something like `settings` that just exports all of these as env variables, for example?

**Comment by cogwirrel** [comment, marked as answer]
Hey! It usually depends a bit on the use case :)

For database DSNs and access tokens I'd do the same of injecting via lambda environment variables in CDK :) Note that for anything secret it's usually best to use secrets manager and pass the secret arn as an environment variable instead of the plaintext value, then retrieve the secret with the AWS SDK in the lambda function itself :)

For an API url - if you want to pass that to say a website you can make use of the `runtimeOptions` in the `StaticWebsite` CDK construct, which lets you pass the value from CDK down to the website when it first loads :)

Hope that helps! :)

**Reply by zetashift** [reply]
That helps a lot, even tho "it depends" is always the answer :P.

I'll inject the DSNs and tokens through variables in CDK.

The API url is a bit more curious tho. I don't use `CDK` here, because I can't find a stable/nice way to get SvelteKit deployed to S3/Lambda.
So I'll just probably use env variables from a global `.env` file or something there.

**Reply by cogwirrel** [reply]
> even tho "it depends" is always the answer :P

So true! 😆 

Ah gotcha - how did you end up deploying SvelteKit? You could always do a similar approach to the `StaticWebsite` construct... `StaticWebsite` writes a json file to the same bucket that the html/js/css is hosted in. I wonder if you could write a json file to wherever your UI is hosted as part of your CDK deployment, such that you can do `fetch('/runtime-config.json')` or similar in your UI? :)

Not a big deal at all (and I imagine could be a bit of work if you're not hosting your UI with infrastructure managed by CDK), but it's always nice to avoid hardcoding this sort of config where possible as it makes it easy to deploy your app to different stages/accounts :)

**Reply by zetashift** [reply]
> how did you end up deploying SvelteKit?

I haven't deployed it yet. A few things aren't going to be static, at this point I'm considering deploying to Vercel or Cloudflare because SvelteKit has official adapters and I already do plenty of hacks to get SvelteKit to play nice with `projen`. 

The downside being using 2 different environments... I'm not sure yet. 

The good news here is that we are planning to use PDK in production! It's been great so far.

>  but it's always nice to avoid hardcoding this sort of config where possible as it makes it easy to deploy your app to different stages/accounts :)

Yes very true! This is what I'm aiming for, but for now it's all done through environment variables.

**Reply by cogwirrel** [reply]
Aha that makes sense!

Great to hear PDK is going well in production! 😁 😁 😁
